File: Facets/View Controllers/Issue/OZLIssueHeaderView.m
//
//  OZLIssueHeaderView.m
//  Facets
//
//  Created by Justin Hill on 11/5/15.
//  Copyright © 2015 Justin Hill. All rights reserved.
//

#import "OZLIssueHeaderView.h"
#import "Facets-Swift.h"

const CGFloat profileSideLen = 32.;
const CGFloat assigneeTextSize = 14.;

@interface OZLIssueHeaderView ()

@property BOOL isFirstLayout;
@property UILabel *assigneeTextLabel;

@end

@implementation OZLIssueHeaderView

#pragma mark - Life cycle
- (instancetype)init {
    if (self = [super init]) {
        self.contentPadding = 16.;
        
        self.titleLabel = [[UILabel alloc] init];
        self.titleLabel.numberOfLines = 0;
        self.titleLabel.lineBreakMode = NSLineBreakByWordWrapping;
        self.titleLabel.font = [UIFont OZLMediumSystemFontOfSize:17];
        
        self.assigneeProfileImageView = [[UIImageView alloc] init];
        self.assigneeProfileImageView.backgroundColor = [UIColor lightGrayColor];
        self.assigneeProfileImageView.layer.cornerRadius = (profileSideLen / 2.);
        self.assigneeProfileImageView.layer.masksToBounds = YES;
        
        self.assigneeTextLabel = [[UILabel alloc] init];
        self.assigneeTextLabel.text = @"ASSIGNEE";
        self.assigneeTextLabel.font = [UIFont systemFontOfSize:10.];
        self.assigneeTextLabel.textColor = [UIColor lightGrayColor];
        
        self.assigneeDisplayNameLabel = [[UILabel alloc] init];
        self.assigneeDisplayNameLabel.font = [UIFont systemFontOfSize:assigneeTextSize];
        
        self.assignButton = [UIButton buttonWithType:UIButtonTypeCustom];
        
        self.isFirstLayout = YES;
    }
    
    return self;
}

#pragma mark - Layout
- (void)layoutSubviews {
    if (self.isFirstLayout) {
        [self addSubview:self.titleLabel];
        [self addSubview:self.assigneeProfileImageView];
        [self addSubview:self.assigneeTextLabel];
        [self addSubview:self.assigneeDisplayNameLabel];
        [self addSubview:self.assignButton];
    }
    
    self.titleLabel.preferredMaxLayoutWidth = self.frame.size.width - (2 * self.contentPadding);
    CGSize size = [self.titleLabel sizeThatFits:CGSizeMake(self.frame.size.width - (2 * self.contentPadding), CGFLOAT_MAX)];
    self.titleLabel.frame = (CGRect){{self.contentPadding, self.contentPadding}, size};
    
    self.assigneeProfileImageView.frame = CGRectMake(self.contentPadding, self.titleLabel.bottom + 12, profileSideLen, profileSideLen);
    
    [self.assigneeTextLabel sizeToFit];
    self.assigneeTextLabel.frame = (CGRect){{self.assigneeProfileImageView.right + 5, self.assigneeProfileImageView.top}, self.assigneeTextLabel.frame.size};
    
    [self.assigneeDisplayNameLabel sizeToFit];
    self.assigneeDisplayNameLabel.frame = (CGRect){{self.assigneeProfileImageView.right + 5, self.assigneeTextLabel.bottom + 3}, self.assigneeDisplayNameLabel.frame.size};
    
    CGRect newAssignFrame = CGRectZero;
    newAssignFrame.origin = self.assigneeProfileImageView.frame.origin;
    newAssignFrame.size.width = MAX(self.assigneeTextLabel.right, self.assigneeDisplayNameLabel.right) - self.assigneeProfileImageView.left;
    newAssignFrame.size.height = self.assigneeDisplayNameLabel.bottom - self.assigneeTextLabel.top;
    self.assignButton.frame = newAssignFrame;
}

- (CGSize)sizeThatFits:(CGSize)size {
    self.frame = (CGRect){CGPointZero, size};
    [self layoutSubviews];
    
    return CGSizeMake(size.width, self.assigneeDisplayNameLabel.bottom + self.contentPadding);
}

#pragma mark - Modeling
- (void)applyIssueModel:(OZLModelIssue *)issue {
    self.titleLabel.attributedText = [self applyTitleAttributesToText:issue.subject];
    
    if (issue.assignedTo) {
        self.assigneeDisplayNameLabel.font = [UIFont systemFontOfSize:assigneeTextSize];
        self.assigneeDisplayNameLabel.text = issue.assignedTo.name;
        self.assigneeDisplayNameLabel.textColor = [UIColor blackColor];
        
    } else {
        self.assigneeDisplayNameLabel.font = [UIFont italicSystemFontOfSize:assigneeTextSize];
        self.assigneeDisplayNameLabel.text = @"Tap to assign";
        self.assigneeDisplayNameLabel.textColor = [UIColor grayColor];
    }
}

- (NSAttributedString *)applyTitleAttributesToText:(NSString *)text {
    NSAssert(text, @"Tried to apply title attributes to nil text");
    
    if (!text) {
        return nil;
    }
    
    NSMutableParagraphStyle *para = [[NSMutableParagraphStyle alloc] init];
    para.lineHeightMultiple = 1.15;
    
    NSAttributedString *attr = [[NSAttributedString alloc] initWithString:text attributes:@{ NSParagraphStyleAttributeName: para} ];
    
    return attr;
}

@end

File: Facets/View Controllers/Issue/OZLIssueHeaderView.swift
//
//  OZLIssueHeaderView.swift
//  Facets
//
//  Created by Justin Hill on 10/29/16.
//  Copyright © 2016 Justin Hill. All rights reserved.
//

import UIKit
import SDWebImage
import Jiramazing

class OZLIssueHeaderView: UIView {
    let profileSideLen: CGFloat = 32.0
    let assigneeTextSize: CGFloat = 14.0
    
    var isFirstLayout = true
    let assigneeTextLabel = UILabel()
    let titleLabel = UILabel()
    let assigneeDisplayNameLabel = UILabel()
    let assigneeProfileImageView = UIImageView()
    let assignButton = UIButton(type: .Custom)
    var contentPadding: CGFloat = 16.0
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        self.titleLabel.numberOfLines = 0
        self.titleLabel.lineBreakMode = .ByWordWrapping
        self.titleLabel.font = UIFont.OZLMediumSystemFontOfSize(17)
        
        self.assigneeProfileImageView.backgroundColor = UIColor.lightGrayColor()
        self.assigneeProfileImageView.layer.cornerRadius = (profileSideLen / 2.0)
        self.assigneeProfileImageView.layer.masksToBounds = true
        
        self.assigneeTextLabel.text = "ASSIGNEE"
        self.assigneeTextLabel.font = UIFont.systemFontOfSize(10)
        self.assigneeTextLabel.textColor = UIColor.lightGrayColor()
        
        self.assigneeDisplayNameLabel.font = UIFont.systemFontOfSize(assigneeTextSize)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        if self.isFirstLayout {
            self.isFirstLayout = false
            self.addSubview(self.titleLabel)
            self.addSubview(self.assigneeProfileImageView)
            self.addSubview(self.assigneeTextLabel)
            self.addSubview(self.assigneeDisplayNameLabel)
            self.addSubview(self.assignButton)
        }
        
        self.titleLabel.preferredMaxLayoutWidth = self.frame.size.width - (2 * self.contentPadding);
        self.titleLabel.frame.origin.x = self.contentPadding
        self.titleLabel.frame.origin.y = self.contentPadding
        self.titleLabel.frame.size = self.titleLabel.sizeThatFits(CGSizeMake(self.frame.size.width - (2 * self.contentPadding), CGFloat.max))
        
        self.assigneeProfileImageView.frame = CGRectMake(self.contentPadding, self.titleLabel.bottom + 12, profileSideLen, profileSideLen);
        
        self.assigneeTextLabel.sizeToFit()
        self.assigneeTextLabel.frame.origin.x = self.assigneeProfileImageView.right + 5
        self.assigneeTextLabel.frame.origin.y = self.assigneeProfileImageView.top
        
        self.assigneeDisplayNameLabel.sizeToFit()
        self.assigneeDisplayNameLabel.frame.origin.x = self.assigneeProfileImageView.right + 5
        self.assigneeDisplayNameLabel.frame.origin.y = self.assigneeTextLabel.bottom + 3
        
        self.assignButton.frame.origin = self.assigneeProfileImageView.frame.origin;
        self.assignButton.frame.size.width = max(self.assigneeTextLabel.right, self.assigneeDisplayNameLabel.right) - self.assigneeProfileImageView.left
        self.assignButton.frame.size.height = self.assigneeDisplayNameLabel.bottom - self.assigneeTextLabel.top
    }
    
    override func sizeThatFits(size: CGSize) -> CGSize {
        self.frame = CGRectMake(0, 0, size.width, size.height)
        self.layoutSubviews()
        
        return CGSizeMake(size.width, self.assigneeDisplayNameLabel.bottom + self.contentPadding);
    }
    
    func applyIssueModel(issue: Issue) {
        if let summary = issue.summary {
            self.titleLabel.attributedText = self.applyTitleAttributesToText(summary)
        } else {
            self.titleLabel.attributedText = nil
        }
        
        if let assignee = issue.assignee {
            self.assigneeDisplayNameLabel.font = UIFont.systemFontOfSize(assigneeTextSize)
            self.assigneeDisplayNameLabel.text = assignee.displayName;
            self.assigneeDisplayNameLabel.textColor = UIColor.blackColor()
            
        } else {
            self.assigneeDisplayNameLabel.font = UIFont.italicSystemFontOfSize(assigneeTextSize)
            self.assigneeDisplayNameLabel.text = "Tap to assign";
            self.assigneeDisplayNameLabel.textColor = UIColor.grayColor()
        }
        
        if let avatarUrl = issue.assignee?.avatarUrls?[.Large] {
            self.assigneeProfileImageView.sd_setImageWithURL(avatarUrl)
        } else {
            self.assigneeProfileImageView.sd_cancelCurrentImageLoad()
            self.assigneeProfileImageView.image = nil
        }
    }
    
    func applyTitleAttributesToText(text: String) -> NSAttributedString {
        let para = NSMutableParagraphStyle()
        para.lineHeightMultiple = 1.15
        
        let attr = NSAttributedString(string: text, attributes: [NSParagraphStyleAttributeName: para])
        
        return attr
    }
}
